x-pack/plugins/apm/server/lib/anomaly_detection/create_anomaly_detection_jobs.ts

```diff
@@ -0,0 +1,123 @@
+  typeof createAnomalyDetectionJobs
+>;
+export async function createAnomalyDetectionJobs(
+  setup: Setup,
+  environments: string[],
+  logger: Logger
+) {
+  const { ml, indices } = setup;
+  if (!ml) {
+    logger.warn('Anomaly detection plugin is not available.');
+    return [];
+  }
+  const mlCapabilities = await ml.mlSystem.mlCapabilities();
+  if (!mlCapabilities.mlFeatureEnabledInSpace) {
+    logger.warn('Anomaly detection feature is not enabled for the space.');
+    return [];
+  }
+  if (!mlCapabilities.isPlatinumOrTrialLicense) {
+    logger.warn(
+      'Unable to create anomaly detection jobs due to insufficient license.'
+    );
+    return [];
+  }
+  logger.info(
+    `Creating ML anomaly detection jobs for environments: [${environments}].`
+  );
+
+  const indexPatternName = indices['apm_oss.transactionIndices'];
+  const responses = await Promise.all(
+    environments.map((environment) =>
+      createAnomalyDetectionJob({ ml, environment, indexPatternName })
+    )
+  );
+  const jobResponses = responses.flatMap((response) => response.jobs);
+  const failedJobs = jobResponses.filter(({ success }) => !success);
+
+  if (failedJobs.length > 0) {
+    const failedJobIds = failedJobs.map(({ id }) => id).join(', ');
+    logger.error(
+      `Failed to create anomaly detection ML jobs for: [${failedJobIds}]:`
+    );
+    failedJobs.forEach(({ error }) => logger.error(JSON.stringify(error)));
+    throw new Error(
+      `Failed to create anomaly detection ML jobs for: [${failedJobIds}].`
+    );
+  }
+
+  return jobResponses;
+}
+
+async function createAnomalyDetectionJob({
+  ml,
+  environment,
+  indexPatternName = 'apm-*-transaction-*',
+}: {
+  ml: Required<Setup>['ml'];
+  environment: string;
+  indexPatternName?: string | undefined;
+}) {
+  const convertedEnvironmentName = convertToMLIdentifier(environment);
+  const randomToken = uuid().substr(-4);
+
+  return ml.modules.setup({
+    moduleId: ML_MODULE_ID_APM_TRANSACTION,
+    prefix: `${ML_GROUP_NAME_APM}-${convertedEnvironmentName}-${randomToken}-`,
+    groups: [ML_GROUP_NAME_APM, convertedEnvironmentName],
+    indexPatternName,
+    query: {
+      bool: {
+        filter: [
+          { term: { [PROCESSOR_EVENT]: 'transaction' } },
+          { exists: { field: TRANSACTION_DURATION } },
+          environment === ENVIRONMENT_NOT_DEFINED
+            ? ENVIRONMENT_NOT_DEFINED_FILTER
+            : { term: { [SERVICE_ENVIRONMENT]: environment } },
+        ],
+      },
+    },
+    startDatafeed: true,
+    jobOverrides: [
+      {
+        custom_settings: {
+          job_tags: { environment },
+        },
+      },
+    ],
+  });
+}
+
+const ENVIRONMENT_NOT_DEFINED_FILTER = {
+  bool: {
+    must_not: {
+      exists: {
+        field: SERVICE_ENVIRONMENT,
+      },
+    },
+  },
+};
+
+export function convertToMLIdentifier(value: string) {
+  return value.replace(/\s+/g, '_').toLowerCase();
```

Comment:
@ogupte we should improve this

```ts
// remove non-alphanumeric characters
retur value.toLowerCase().replace(/[^a-z0-9+]+/g, '_');
```

Reply:
i simplified this in another PR: https://github.com/elastic/kibana/blob/55b1e911228f55aa778c749d74999990d8339aa8/x-pack/plugins/apm/server/lib/anomaly_detection/create_anomaly_detection_jobs.ts#L86 to just use the lodash function `snakeCase`. This also has the effect of deburring and removing all invalid chars